src/ut/httpconnection_test.cpp

```diff
@@ -105,14 +108,124 @@ class HttpConnectionTest : public BaseTest
+
+    fakecurl_responses["http://3.0.0.0:80/tcp_success"] = CURLE_REMOTE_FILE_NOT_FOUND;
+
+    fakecurl_responses["http://3.0.0.0:80/one_failure"] = CURLE_COULDNT_RESOLVE_HOST;
+    fakecurl_responses["http://3.0.0.1:80/one_failure"] = "<message>success</message>";
+
+    fakecurl_responses["http://3.0.0.0:80/all_failure"] = CURLE_COULDNT_RESOLVE_HOST;
+    fakecurl_responses["http://3.0.0.1:80/all_failure"] = CURLE_COULDNT_RESOLVE_HOST;
+  }
+
+  ~HttpConnectionBlacklistTest()
+  {
+    fakecurl_responses.clear();
+    fakecurl_requests.clear();
+    delete _http; _http = NULL;
+    delete _cm; _cm = NULL;
+    delete _am; _am = NULL;
+  }
+
+  /// Creates a vector of count AddrInfo targets, beginning from 3.0.0.0 and
+  /// incrementing by one each time.
+  std::vector<AddrInfo> create_targets(int count)
+  {
+    std::vector<AddrInfo> targets;
+    AddrInfo ai;
+    ai.port = 80;
+    ai.transport = IPPROTO_TCP;
+    std::stringstream os;
+    for (int i = 0; i < count; ++i)
+    {
+      os << "3.0.0." << i;
+      BaseResolver::parse_ip_target(os.str(), ai.address);
+      targets.push_back(ai);
+      os.str(std::string());
+    }
+    return targets;
+  }
+};
+
+TEST_F(HttpConnectionBlacklistTest, BlacklistTestHttpSuccess)
+{
+  std::vector<AddrInfo> targets = create_targets(2);
+
+  EXPECT_CALL(_resolver, resolve(_,_,_,_,_)).WillOnce(SetArgReferee<3>(targets));
+  EXPECT_CALL(_resolver, success(targets[0])).Times(1);
+  EXPECT_CALL(_resolver, untested(targets[1])).Times(1);
+
+  string output;
+  _http->send_get("/http_success", output, "", 0);
+}
+
+TEST_F(HttpConnectionBlacklistTest, BlacklistTestTcpSuccess)
+{
+  std::vector<AddrInfo> targets = create_targets(2);
+
+  EXPECT_CALL(_resolver, resolve(_,_,_,_,_)).WillOnce(SetArgReferee<3>(targets));
+  EXPECT_CALL(_resolver, success(targets[0])).Times(1);
+  EXPECT_CALL(_resolver, untested(targets[1])).Times(1);
+
+  string output;
+  _http->send_get("/tcp_success", output, "", 0);
+}
+
+TEST_F(HttpConnectionBlacklistTest, BlacklistTestOneFailure)
+{
+  std::vector<AddrInfo> targets = create_targets(2);
+
+  EXPECT_CALL(_resolver, resolve(_,_,_,_,_)).WillOnce(SetArgReferee<3>(targets));
+  EXPECT_CALL(_resolver, blacklist(targets[0])).Times(1);
+  EXPECT_CALL(_resolver, success(targets[1])).Times(1);
+
+  string output;
+  _http->send_get("/one_failure", output, "", 0);
+}
+
+TEST_F(HttpConnectionBlacklistTest, BlacklistTestAllFailure)
+{
+  std::vector<AddrInfo> targets = create_targets(2);
+
+  EXPECT_CALL(_resolver, resolve(_,_,_,_,_)).WillOnce(SetArgReferee<3>(targets));
+  EXPECT_CALL(_resolver, blacklist(targets[0])).Times(1);
+  EXPECT_CALL(_resolver, blacklist(targets[1])).Times(1);
+
+  string output;
+  _http->send_get("/all_failure", output, "", 0);
+}
 
 TEST_F(HttpConnectionTest, SimpleKeyAuthGet)
 {
   string output;
   long ret = _http->send_get("/blah/blah/blah", output, "gandalf", 0);
+
```

Comment:
OK, so I think there are actually too many blank lines here. I realize this is contradictory advice so let's discuss on Monday.